File: src/leetcode/python/MinimumIndexSumOfTwoLists_599.py
class MinimumIndexSumOfTwoLists_599(object):
    def findRestaurant(self, list1, list2):
        """
        :type list1: List[str]
        :type list2: List[str]
        :rtype: List[str]
        """
        d1 = {v: i for i, v in enumerate(list1)}
        d2 = {v: i for i, v in enumerate(list2)}
        com = set(list1) & set(list2)
        res = []
        m = 1e9
        comset = {}
        for s in com:
            v = d1[s] + d2[s]
            m = min(m, v)
            comset[s] = v
        for k in comset.keys():
            if comset[k] == m:
                res.append(k)
        return res

File: src/leetcode/java/MinimumIndexSumOfTwoLists_599.java
package leetcode.java;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MinimumIndexSumOfTwoLists_599 {
  public String[] findRestaurant(String[] list1, String[] list2) {
    Map<String, Integer> l1 = new HashMap<>();
    Map<String, Integer> common = new HashMap<>();
    for (int i = 0; i < list1.length; i++) l1.put(list1[i], i);
    int min = Integer.MAX_VALUE;
    int val = 0;
    for (int j = 0; j < list2.length; j++) {
      if (l1.containsKey(list2[j])) {
        val = j + l1.get(list2[j]);
        min = Math.min(min, val);
        common.put(list2[j], val);
      }
    }
    List<String> c = new ArrayList<>();
    for (Map.Entry<String, Integer> entry : common.entrySet()) {
      if (entry.getValue() == min) c.add(entry.getKey());
    }
    return c.toArray(new String[c.size()]);
  }
}
